### PR TITLE
fix(core): preserve baseline centering in Hedge updates when masked-in mass underflows

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2409

## Summary
In Hedge-style preference updates (`LearnedResourceManager`, `GeneratorMetaResourceManager`, and `FixedBudgetFeatureLearner`), centering baselines were computed by dividing masked weights by `jnp.maximum(jnp.sum(jnp.where(mask, weights, 0.0)), 1e-12)`. When masked-in weights carry less mass than `1e-12` (e.g. after a large preference gap where the leader is masked out by a NaN), the divisor clamped to `1e-12`, causing `masked_weights` to sum to $< 1.0$ (or $0.0$) and pulling the baseline towards $0$ rather than the participating actions, thus silently losing centering.

## Proposed Changes
1. Added `_masked_normalized_weights` to normalize active weights over the valid mask and fall back to uniform distribution over the mask when valid mass underflows or is zero.
2. Applied `_masked_normalized_weights` across `LearnedResourceManager._update_jit`, `GeneratorMetaResourceManager._update_jit`, and `FixedBudgetFeatureLearner._resource_log_weight_update`.
3. Added unit and regression test in `tests/test_resource_manager.py`.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_resource_manager.py tests/test_feature_discovery_loop_steps.py` (79/79 passed).
- Ran ruff: `/opt/homebrew/bin/uv run ruff check alberta_framework/core/resource_manager.py alberta_framework/core/feature_discovery.py tests/test_resource_manager.py` (clean).
- Ran mypy: `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/core/resource_manager.py alberta_framework/core/feature_discovery.py` (clean).

### CODE REVIEW

**Review by mashingaan — APPROVED** [review]
## Independent review

Verified against exact head 6c17754920cab3e45ee15d96ca6337ed01795d4a.

The fix addresses issue #2409 by normalizing only the active masked weights and falling back to a uniform active distribution when the mass is zero or no longer representable. This restores the simplex invariant and preserves centered advantages under extreme preference gaps across the three affected update paths.

Independent verification in a disposable exact-head checkout:

- uv run --extra dev python -m pytest -q tests/test_resource_manager.py -> 67 passed
- uv run --extra dev python -m pytest -q tests/test_feature_discovery_loop_steps.py -> 12 passed
- Ruff passed on the changed production files
- An underflow probe returned [0, 0.5, 0.5] with total 1.0
- Exact-head GitHub CI checks were successful

The review is bound to commit 6c17754920cab3e45ee15d96ca6337ed01795d4a. Approve.

Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
AI provider/model: openai / gpt-5.6-sol
Client / agent tooling: codex
Contribution skill revision: SlopDotCash/slopdotcash@1a2017b9fbbcef4263da9dda8d0b4e1d0a7bff42:skills/contribute-to-asi
Attribution status: self-reported
— [asi-pr2494-review]

**Review by kenjohnscreates — CHANGES_REQUESTED** [review]
## Disposition: **request changes** (the `1e-12` floor survives as the validity threshold, so unequal survivors below it get a uniform baseline)

Exact head reviewed: `6c177549` on base `c9aba7b5`. Filed 2026-08-25 17:06 UTC.

The equal-survivor rows and the shift rows are repaired (`+1,-1` at every gap and shift), and the well-formed sweep and trajectory hash are byte-identical to main. Touched test files in this head's worktree (`test_resource_manager.py`, `test_update_safety.py`, `test_feature_discovery.py`): 233 passed; ruff clean.

**Blocking, measured.** `_masked_normalized_weights` divides by `max(mass, 1e-12)` and accepts the result only when `mass > 1e-12`; otherwise it returns uniform over the mask. That reproduces the issue's threshold rather than removing it. The issue's table uses equal survivors, where uniform and the true convex combination coincide, so the new test cannot see this; with unequal survivors it is wrong:

| `log_weights` | mass | true convex advantages | this head |
|---|---|---|---|
| `[28,0,-1]` | `1.4e-12` | `+0.5379,-1.4621` | `+0.5379,-1.4621` |
| `[30,0,-1]` | `1.9e-13` | `+0.5379,-1.4621` | **`+1,-1`** |
| `[40,0,-1]`, `[60,0,-1]` | `8.5e-18`, `1.8e-26` | `+0.5379,-1.4621` | **`+1,-1`** |
| `[40,0,-5]` | `8.5e-18` | `+0.0134,-1.9866` | **`+1,-1`** |
| `GeneratorMetaResourceManager` `[40,0,-1]` | | `-0.5379,+1.4621` | **`-1,+1`** |
| `FixedBudget` allocation `[1,3e-13,1e-13]`, mask `[F,T,T]` | `4e-13` | convex | **uniform** |

Every other head in the cluster gets these rows right because it divides by the actual mass whenever it is positive (or normal). Fix: drop the `1e-12` from both the divisor and the validity test, gate on `mass > 0` (or `mass >= finfo.tiny`) and check the outcome, as #2410 does. Also measured: NaN or inf inside the mask is replaced by a uniform row here, where main returned NaN and #2410 passes it through. mashingaan's approval on this head did not drive an unequal-survivor row.

Under the cluster tie-break the approval goes to #2410.

## Measured across the cluster

| Probe row | main | #2410 | #2494 | the other eight heads |
|---|---|---|---|---|
| equal survivors `[gap,0,0]`, gaps 30 ... 200 (want `+1,-1`) | `-0.63,-2.63` ... `-1,-3` | `+1,-1` at every gap | same | same |
| uniform shift 1 / 10 / 100 at gap 40 (want `+1,-1`) | `-2,-4`; `-10,-10`; `-10,-10` | `+1,-1` | same | same |
| unequal survivors `[gap,0,-1]`, gaps 28 ... 60 (want `+0.5379,-1.4621`) | `+0.45` ... `-1,-3` | `+0.5379,-1.4621` | **`+1,-1` (uniform)** | `+0.5379,-1.4621` |
| `[gap,0,-5]` at gap 40 (want `+0.0134,-1.9866`) | `-1,-3` | `+0.0134,-1.9866` | **`+1,-1`** | `+0.0134,-1.9866` |
| `[gap,0,-5]` at gap 85 (weight of the `-5` survivor flushes) | `-1,-3` | `0,-2` (sole representable survivor) | `+1,-1` | `0,-2` |
| `[gap,0,-5]` at gaps 90 ... 200 (both survivors flush; conditional softmax would be `+0.0134,-1.9866`) | `-1,-3` | `+1,-1` (uniform) | `+1,-1` | `+1,-1` |
| single survivor at gap 200 (want `0`) | `-1` | `0` | `0` | `0` |
| `GeneratorMetaResourceManager` equal survivors gap 40 / 200 (want `-1,+1`) | `+1,+3` | `-1,+1` | same | same |
| `GeneratorMetaResourceManager` unequal `[40,0,-1]` (want `-0.5379,+1.4621`) | `+1,+3` | correct | **`-1,+1`** | correct |
| `FixedBudget` allocation `[1, 3e-13, 1e-13]` mask `[F,T,T]` | not convex | convex (`-0.33,-0.83,+1.17`) | **uniform** | convex |
| `FixedBudget` `finfo.tiny` pair / `tiny/2` pair / zero mass | not convex | uniform over the mask | uniform | uniform |
| `FixedBudget` NaN or inf inside the mask | NaN | **NaN passed through** | uniform | uniform (#2428, #2461: NaN -> uniform, inf -> NaN) |
| 2000-row well-formed sweep (adv + new log-weights) | `483889a0ffce` / `2fb8e40484a1` | **identical** | identical | identical |
| 300-step trajectory hash | `19a74538c209` | **identical** | identical | identical |

Every head is byte-identical to main wherever the participating mass was representable above the old floor, repairs the equal-survivor rows and the shift rows, and keeps `update_applied` true. The one construction that fails a defect row is #2494, which keeps `1e-12` as its validity threshold and falls back to uniform below it; on unequal survivors that is the wrong convex combination. Shared by all ten and raised by saltwire-dev on #2461: every head renormalizes the leftover linear-space softmax mass, so once the survivors' own weights flush (gap above ~88 nats for the smaller survivor) the fallback is uniform, whereas a masked softmax in log space (`softmax(where(valid, log_weights, -inf))`) would still recover the conditional distribution; the issue text accepts uniform in that regime, so it is a follow-up, not a discriminator. Every head's touched tests pass and `ruff` is clean.


---
Reviewed with claude-code / anthropic / claude-fable-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

**Comment by lalalune** [comment]
Closing as superseded by merged #2526. The merged power-of-two rescaling keeps the baseline a convex combination over participating allocations without denominator floors or reciprocal loss of finite ratios. The live regression failed before the fix and all 215 resource-manager/feature-discovery tests pass.
